content/lectures/lec03.md

```diff
@@ -7,4 +7,192 @@ slides: ""
+```
+
+### Dictionaries
+
+Use two spaces in the indent and YAML will associate the values as a dictionary.
+```yaml
+# 2 spaces per indent or bust
+staff_ages:
+  instructors:
+    armaan: 21
+    peyton: 20
+  tas:
+    davis: 22
+    campbell: 20
+```
+The example above has a readability problem. What is the value that we are associating with the names? We see that Peyton is 21 and Davis is 22, but what does that mean? Are they 21 and 22 feet tall? The example below shows how we can use lists of dictionaries to remove that ambiguity:
+```yaml
+staff:
+  instructors:
+    - name: armaan
+      age: 21
+    - name: peyton
+      age: 20
+  tas:
+    - name: davis
+      age: 22
+    - name: campbell
+      age: 20
+```
+
+## YAML Lint
+
+Check out [yamllint.com/](http://www.yamllint.com/) to double check your YAML, it will help remove ambiguity from the config. We automatically run this in the autograder so definitely put your Docker compose YAML in here before you submit!
+
+## Docker-Compose to Production?
+
+So far, with all of these Docker tools (Docker images, Docker containers, Docker runtime, Docker compose, and Docker Hub) we can build and distribute images, and coordinate multiple containers together. If we're designing a production system that we want to serve actual web traffic, is this enough? Are we missing anything? The answer is yes! Here are four main pillars that are not handled at all by docker-compose: 
+
+- High Availability: If a node fails, will I still be able to serve traffic?
+- Horizontal Scalability: We can duplicate Docker containers, but still we are limited by the hardware capacity of the single node that is running Docker. Also, we cannot scale automatically based on the amount of traffic we are receiving.
+  - Vertical scalability is when we duplicate a service many times on the same machine, Docker compose allows us to do this by duplicating containers. Horizontal scalability is when we duplicate our service many times across different machines. We want horizontal scalability so that we are not limited by the hardware of a single machine. Horizontal scalability also affords us availability, if only one of the many machines running our code fails, we can still serve traffic from the other machines.
+- Networking: Docker compose allows us to network between containers on a single machine, but we cannot communicate across multiple machines. If we want scalability and availability, we're going to need to communicate between multiple machines.
+- Observability: How can we monitor all of our containers and their health? Are we notified if something fails? Are we getting more traffic than we can handle?
+
+Note that all of these pillars come with tradeoffs. We can have super high availability, but it might make networking or observability more difficult. There are going to be tradeoffs and it won't always be the case that one technology is the solution to every existing use case.
```

Comment:
The tradeoffs we talk about are less about conflicts between these goals and more about conflict between these goals and system complexity

Reply:
the paragraph below does a really good job of talking about this point. i'd just delete this paragraph